builder/etcd/etcd.go

```diff
@@ -0,0 +1,530 @@
+	out, err := exec.Command("ssh-keygen", "-A").CombinedOutput()
+	if err != nil {
+		log.Infof(c, "ssh-keygen: %s", out)
+		log.Errf(c, "Failed to generate SSH keys: %s", err)
+		return err
+	}
+	return nil
+}
+
+// UpdateHostPort intermittently notifies etcd of the builder's address.
+//
+// If `port` is specified, this will notify etcd at 10 second intervals that
+// the builder is listening at $HOST:$PORT, setting the TTL to 20 seconds.
+//
+// This will notify etcd as long as the local sshd is running.
+//
+// Params:
+// 	- base (string): The base path to write the data: $base/host and $base/port.
+// 	- host (string): The hostname
+// 	- port (string): The port
+// 	- client (Setter): The client to use to write the data to etcd.
+// 	- sshPid (int): The PID for SSHD. If SSHD dies, this stops notifying.
+func UpdateHostPort(c cookoo.Context, p *cookoo.Params) (interface{}, cookoo.Interrupt) {
+
+	base := p.Get("base", "").(string)
+	host := p.Get("host", "").(string)
+	port := p.Get("port", "").(string)
+	client := p.Get("client", nil).(Setter)
+	sshd := p.Get("sshdPid", 0).(int)
+
+	// If no port is specified, we don't do anything.
+	if len(port) == 0 {
+		log.Infof(c, "No external port provided. Not publishing details.")
+		return false, nil
+	}
+
+	var ttl uint64 = 20
+
+	if err := setHostPort(client, base, host, port, ttl); err != nil {
+		log.Errf(c, "Etcd error setting host/port: %s", err)
+		return false, err
+	}
+
+	// Update etcd every ten seconds with this builder's host/port.
+	safely.GoDo(c, func() {
+		ticker := time.NewTicker(10 * time.Second)
+		for range ticker.C {
+			//log.Infof(c, "Setting SSHD host/port")
+			if _, err := os.FindProcess(sshd); err != nil {
+				log.Errf(c, "Lost SSHd process: %s", err)
+				break
+			} else {
+				if err := setHostPort(client, base, host, port, ttl); err != nil {
+					log.Errf(c, "Etcd error setting host/port: %s", err)
+					break
+				}
+			}
+		}
+		ticker.Stop()
+	})
+
+	return true, nil
+}
+
+func setHostPort(client Setter, base, host, port string, ttl uint64) error {
+	if _, err := client.Set(base+"/host", host, ttl); err != nil {
+		return err
+	}
+	if _, err := client.Set(base+"/port", port, ttl); err != nil {
+		return err
+	}
+	return nil
+}
+
+// MakeDir makes a directory in Etcd.
+//
+// Params:
+// 	- client (EtcdDirCreator): Etcd client
+//  - path (string): The name of the directory to create.
+// 	- ttl (uint64): Time to live.
+// Returns:
+// 	*etcd.Response
+func MakeDir(c cookoo.Context, p *cookoo.Params) (interface{}, cookoo.Interrupt) {
+	name := p.Get("path", "").(string)
+	ttl := p.Get("ttl", uint64(0)).(uint64)
+	cli, ok := p.Has("client")
+	if !ok {
+		return nil, errors.New("No Etcd client found.")
+	}
+	client := cli.(DirCreator)
+
+	if len(name) == 0 {
+		return false, errors.New("Expected directory name to be more than zero characters.")
+	}
+
+	res, err := client.CreateDir(name, ttl)
+	if err != nil {
+		// FIXME: Do we want this to be recoverable?
```

Comment:
I think this is fine for now. As long as we log the error, the user can debug from there and check what's happening with etcd.